control-plane/api-gateway/controllers/gatewayclass_controller.go

```diff
@@ -0,0 +1,264 @@
+}
+
+// SetupWithManager registers the controller with the given manager.
+func (r *GatewayClassController) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&gwv1beta1.GatewayClass{}).
+		// Watch for changes to GatewayClassConfig objects.
+		Watches(source.NewKindWithCache(&v1alpha1.GatewayClassConfig{}, mgr.GetCache()), r.gatewayClassConfigFieldIndexEventHandler(ctx)).
+		// Watch for changes to Gateway objects that reference this GatewayClass.
+		Watches(source.NewKindWithCache(&gwv1beta1.Gateway{}, mgr.GetCache()), r.gatewayFieldIndexEventHandler(ctx)).
+		Complete(r)
+}
+
+// isGatewayClassInUse returns true if the given GatewayClass is referenced by any Gateway objects.
+func (r *GatewayClassController) isGatewayClassInUse(ctx context.Context, gc *gwv1beta1.GatewayClass) (bool, error) {
+	list := &gwv1beta1.GatewayList{}
+	if err := r.Client.List(ctx, list, &client.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector(Gateway_GatewayClassIndex, gc.Name),
+	}); err != nil {
+		return false, err
+	}
+
+	return len(list.Items) != 0, nil
+}
+
+// validateParametersRef validates the ParametersRef field of the given GatewayClass
+// if it is set, ensuring that the referenced object is a GatewayClassConfig that exists.
+func (r *GatewayClassController) validateParametersRef(ctx context.Context, gc *gwv1beta1.GatewayClass, log logr.Logger) (didUpdate bool, err error) {
+	parametersRef := gc.Spec.ParametersRef
+	if parametersRef == nil {
+		return false, nil
+	}
+
+	if parametersRef.Kind != v1alpha1.GatewayClassConfigKind {
+		didUpdate = r.setCondition(gc, metav1.Condition{
+			Type:    invalidParameters,
+			Status:  metav1.ConditionTrue,
+			Reason:  configurationInvalid,
+			Message: fmt.Sprintf("Incorrect type for parametersRef. Expected GatewayClassConfig, got %q.", parametersRef.Kind),
+		})
+		return didUpdate, nil
+	}
+
+	err = r.Client.Get(ctx, types.NamespacedName{Name: parametersRef.Name}, &v1alpha1.GatewayClassConfig{})
+	if k8serrors.IsNotFound(err) {
+		didUpdate := r.setCondition(gc, metav1.Condition{
+			Type:    invalidParameters,
+			Status:  metav1.ConditionTrue,
+			Reason:  configurationInvalid,
+			Message: fmt.Sprintf("GatewayClassConfig not found %q.", parametersRef.Name),
+		})
+		return didUpdate, err
+	}
+	if err != nil {
+		log.Error(err, "unable to fetch GatewayClassConfig")
+	}
+	return false, err
+}
+
+// setCondition sets the given condition on the given GatewayClass.
+func (r *GatewayClassController) setCondition(gc *gwv1beta1.GatewayClass, condition metav1.Condition) (didUpdate bool) {
+	condition.LastTransitionTime = metav1.Now()
+
+	// Set the condition if it already exists.
+	for i, c := range gc.Status.Conditions {
+		if c.Type == condition.Type {
+			// The condition already exists and is up to date.
+			if equalConditions(condition, c) {
+				return false
+			}
+
+			condition.ObservedGeneration = c.ObservedGeneration + 1
```

Comment:
The generation comes from the object that's being observed: https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#ObjectMeta.GetGeneration

What that means is that we should be doing `condition.ObservedGeneration = gc.GetGeneration()` and then using that as part of the equality check above (the generation value won't change on the object unless the spec has been updated, so it should be fine to do the status update and not worry about infinite loops even if the generation is part of the equality check)

Reply:
Great! Added!